code-studio/build-js.js

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env node
+/** @file Build script for JS assets in the code-studio package, which is loaded
+    by dashboard (our "Code Studio" Rails app). */
+'use strict';
+
+var _ = require('lodash');
+var child_process = require('child_process');
+var program = require('commander');
+var path = require('path');
+
+/** @const {string} */
+var SRC_PATH = './src/js/';
+
+/** @const {string} */
+var BUILD_PATH = './build/js/';
+
+/**
+ * Files to build, given as paths rooted at code-studio/src/js/
+ * Each will result in an output file.
+ * @type {string[]}
+ */
+var FILES = [
+  'levelbuilder.js',
+  'levelbuilder_dsl.js',
+  'levelbuilder_studio.js',
+];
+
+/**
+ * Build script main entry point.
+ */
+function main() {
+  // Use commander to parse command line arguments
+  // https://github.com/tj/commander.js
+  program
+      .option('--min', 'Build minified output', false)
+      .parse(process.argv);
+
+  // Run build
+  var buildCommands = [
+    ensureDirectoryExists(BUILD_PATH),
+    browserifyCommand(program.min)
+  ].join(" && \\\n");
+  console.log(buildCommands);
+
+  try {
+    // For documentation on synchronous execution of shell commands from node scripts, see:
+    // https://nodejs.org/docs/latest/api/child_process.html#child_process_synchronous_process_creation
+    var result = child_process.execSync(buildCommands, {
+      env: _.extend({}, process.env, {
+        PATH: './node_modules/.bin:' + process.env.PATH
+      }),
+      stdio: 'inherit'
+    });
+    console.log("code-studio built\n");
+  } catch (e) {
+    process.exit(e.status);
+  }
+}
+
+/**
+ * Generate command to:
+ * Creates given director(y|ies) to any depth if it does not exist, no-op if
+ * it does already exist.
+ * @param {!string} dir
+ * @returns {string}
+ */
+function ensureDirectoryExists(dir) {
+  return 'mkdir -p ' + dir;
+}
+
+/**
+ * Generate command to:
+ * Bundle JavaScript files using Browserify, break out common code using
+ * factor-bundle, and (optionally) minify output using uglify.
+ * @param {boolean} [shouldMinify]
+ * @returns {string}
+ */
+function browserifyCommand(shouldMinify) {
```

Comment:
What do you think of passing in FILES, BUILD_PATH, and SRC_PATH rather than depending on module level variables? This way if we want to extract this method elsewhere for some reason, it becomes a lot easier :)